command/ca/policy/policycontext/context.go

```diff
@@ -0,0 +1,186 @@
+	return isPolicyLevel(ctx, provisionerPolicyLevel)
+}
+
+// WithACMEPolicyLevel returns a new context.Context with
+// parent ctx and ACME account policy level set.
+func WithACMEPolicyLevel(ctx context.Context) context.Context {
+	return context.WithValue(ctx, policyLevelContextKey{}, acmePolicyLevel)
+}
+
+// IsACMEPolicyLevel returns if the context.Context has ACME account policy level.
+func IsACMEPolicyLevel(ctx context.Context) bool {
+	return isPolicyLevel(ctx, acmePolicyLevel)
+}
+
+// isPolicyLevel checks if the context.Context has the specified policy level set.
+func isPolicyLevel(ctx context.Context, level policyLevel) bool {
+	if ctx == nil {
+		return false
+	}
+	v, _ := ctx.Value(policyLevelContextKey{}).(policyLevel)
+	return v == level
+}
+
+type policyConfigurationTypeContextKey struct{}
+
+type policyConfigurationType int
+
+const (
+	x509PolicyType policyConfigurationType = iota + 1
+	sshHostPolicyType
+	sshUserPolicyType
+)
+
+// WithX509Policy returns a new context.Context with
+// parent ctx and X509 policy set.
+func WithX509Policy(ctx context.Context) context.Context {
+	return context.WithValue(ctx, policyConfigurationTypeContextKey{}, x509PolicyType)
+}
+
+// IsX509Policy returns if the context.Context has X.509 policy set.
+func IsX509Policy(ctx context.Context) bool {
+	if ctx == nil {
+		return false
+	}
+	v, _ := ctx.Value(policyConfigurationTypeContextKey{}).(policyConfigurationType)
+	return v == x509PolicyType
+}
+
+// WithSSHHostPolicy returns a context.Context with SSH host policy set.
+func WithSSHHostPolicy(ctx context.Context) context.Context {
+	return context.WithValue(ctx, policyConfigurationTypeContextKey{}, sshHostPolicyType)
+}
+
+// IsSSHHostPolicy returns if the context.Context has SSH host policy set.
+func IsSSHHostPolicy(ctx context.Context) bool {
+	if ctx == nil {
+		return false
+	}
+	v, _ := ctx.Value(policyConfigurationTypeContextKey{}).(policyConfigurationType)
+	return v == sshHostPolicyType
+}
+
+// WithSSHUserPolicy returns a context.Context with SSH user policy set.
+func WithSSHUserPolicy(ctx context.Context) context.Context {
+	return context.WithValue(ctx, policyConfigurationTypeContextKey{}, sshUserPolicyType)
+}
+
+// IsSSHUserPolicy returns if context.Context has SSH user policy set.
+func IsSSHUserPolicy(ctx context.Context) bool {
+	if ctx == nil {
+		return false
+	}
+	v, _ := ctx.Value(policyConfigurationTypeContextKey{}).(policyConfigurationType)
+	return v == sshUserPolicyType
+}
+
+type policyTypeContextKey struct{}
+
+type policyType int
+
+const (
+	allowType policyType = iota + 1
+	denyType
+)
+
+// WithAllow returns a context.Context with allow policy set.
+func WithAllow(ctx context.Context) context.Context {
+	return context.WithValue(ctx, policyTypeContextKey{}, allowType)
+}
+
+// IsAllow returns if the context.Context has allow set.
+func IsAllow(ctx context.Context) bool {
+	if ctx == nil {
+		return false
+	}
```

Comment:
Perhaps we should avoid sending nil here.

Reply:
The only way to fully prevent that is to check that no-one sending a nil here, either through static analysis or code review. I've chosen to `panic` now, since I consider this a programmer error if it happens.

Reply:
As I commented, a panic will also happen without that condition, so I would remove it.